editor/code_editor.cpp

```diff
@@ -1418,48 +1418,68 @@ void CodeTextEditor::delete_lines() {
 	text_editor->end_complex_operation();
 }
 
-void CodeTextEditor::duplicate_selection() {
+void CodeTextEditor::duplicate_lines() {
 	text_editor->begin_complex_operation();
 
 	Vector<int> caret_edit_order = text_editor->get_caret_index_edit_order();
-	for (const int &c : caret_edit_order) {
-		const int cursor_column = text_editor->get_caret_column(c);
-		int from_line = text_editor->get_caret_line(c);
-		int to_line = text_editor->get_caret_line(c);
-		int from_column = 0;
-		int to_column = 0;
-		int cursor_new_line = to_line + 1;
-		int cursor_new_column = text_editor->get_caret_column(c);
-		String new_text = "\n" + text_editor->get_line(from_line);
-		bool selection_active = false;
-
-		text_editor->set_caret_column(text_editor->get_line(from_line).length(), c == 0, c);
-		if (text_editor->has_selection(c)) {
-			from_column = text_editor->get_selection_from_column(c);
-			to_column = text_editor->get_selection_to_column(c);
-
-			from_line = text_editor->get_selection_from_line(c);
-			to_line = text_editor->get_selection_to_line(c);
-			cursor_new_line = to_line + text_editor->get_caret_line(c) - from_line;
-			cursor_new_column = to_column == cursor_column ? 2 * to_column - from_column : to_column;
-			new_text = text_editor->get_selected_text(c);
-			selection_active = true;
-
-			text_editor->set_caret_line(to_line, c == 0, true, 0, c);
-			text_editor->set_caret_column(to_column, c == 0, c);
-		}
+	for (const int &caret_index : caret_edit_order) {
+		// The text that will be inserted. All lines in one string.
+		String insert_text;
+
+		// The new line position of the caret after the operation.
+		int new_caret_line = text_editor->get_caret_line(caret_index);
+		// The new column position of the caret after the operation.
+		int new_caret_column = text_editor->get_caret_column(caret_index);
+		// The caret positions of the selection. Stays -1 if there is no selection.
+		int select_from_line = -1;
+		int select_to_line = -1;
+		int select_from_column = -1;
+		int select_to_column = -1;
+		// Number of lines of the selection.
+		int select_num_lines = -1;
+
+		if (text_editor->has_selection(caret_index)) {
+			select_from_line = text_editor->get_selection_from_line(caret_index);
+			select_to_line = text_editor->get_selection_to_line(caret_index);
+			select_from_column = text_editor->get_selection_from_column(caret_index);
+			select_to_column = text_editor->get_selection_to_column(caret_index);
+			select_num_lines = select_to_line - select_from_line + 1;
+
+			for (int i = select_from_line; i <= select_to_line; i++) {
+				insert_text += "\n" + text_editor->get_line(i);
+			}
+			new_caret_line = select_to_line;
+
+			new_caret_line += select_num_lines;
```

Comment:
`new_caret_line = select_to_line + select_num_lines;`

Reply:
You are right. This has been overlooked since removing the duplicate_lines_above code.